dev_docs/tutorials/versioning_interfaces.mdx

```diff
@@ -0,0 +1,418 @@
+date: 2023-02-09
+tags: ['kibana', 'onboarding', 'dev', 'architecture']
+---
+
+To support versioned APIs we need to keep past versions of interfaces around. This tutorial presents one strategy to manage your interfaces.
+
+## The strategy
+
+Every plugin has a _domain_. A domain consists of a one or more concepts, usually objects, that are related in some logical way.
+
+At a high level the strategy for versioning our interfaces is:
+
+> Version a collection of related interfaces together. Whenever a single interface changes, increment the version of the entire collection of interfaces.
+
+Characteristics of this strategy:
+
+1. Avoid the `extends` keyword for referencing past interfaces. This results in cleaner API docs being generated.
+2. Leverage `* as` to create versioned namespaces rather than versioned interfaces.
+3. Verbosity is intentional: for a single change in a collection, we create a new version of the related collection.
+
+### An example: Weather insights unversioned
+Consider a fictional Kibana plugin called "Weather insights". It allows users to easily create and curate dashboards that house visualisations and metrics for weather data. The domain contains the following:
+
+1. **Data source**: an endpoint that is scraped and ingested into ES for aggregation and search.
+2. **Weather dashboard**: a set of visualisations and metrics that draws from a data source.
+
+We have the following `common` interfaces:
+
+```ts
+interface DataSource {
+  /** Some URL that provides weather data in an expected format for scraping */
+  url: string;
+  /** The data view data is stored in */
+  dataViewId: string;
+  /** Username to use with source URL */
+  username?: string;
+  /** Password to use with source URL */
+  password?: string;
+}
+
+/** Fictional type, only used for this example as an external type */
+import type { PortableDashboard } from '@kbn/dashboard/common';
+
+interface WeatherDashboard {
+  /** A unqiue ID for this weather dashboard */
+  id: string;
+  /** Definition of a dashboard imported from the `dashboard` plugin */
+  dashboard: PortableDashboard;
+  /** Data source to use with this dashboard */
+  source: DataSource;
+}
+
+type UnsavedWeatherDashboard = Omit<WeatherDashboard, 'id'>;
+
+/** Below are the HTTP APIs based on the domain types */
+
+interface GetWeatherDashboardHTTPResponse {
+  weatherDashboard: WeatherDashboard;
+}
+
+interface CreateWeatherDashboardHTTPBody {
+  weatherDashboard: UnsavedWeatherDashboard;
+}
+
+interface CreateWeatherDashboardHTTPResponse {
+  weatherDashboard: WeatherDashboard;
+}
+
+// Imagine this pattern for the update and list endpoints too.
+```
+
+### An example: Weather insights versioned
+
+The following product requirements are added to the Weather insights plugin:
+
+1. Data sources need a more user-friendly name, currently we just show a URL
```

Comment:
nit: if name is optional this could be a bwc change and we don't need a v2. Maybe we should add: "All new data sources _must_ specify a name field."

Wondering if we should say "Existing documents without a name field should populate the name field with the URL"... maybe not super relevant to the subject and interfaces, but maybe provides a bit of color.

Reply:
I think what we're trying to say is that: "There is a requirement for data sources to have a human-readable `name` property rather than a `URL`".
I could be wrong though.

Reply:
That is true, how about we add an optional field `description` and add a comment that if we only added `description` there would be no need for a new route version.